### PR TITLE
API-1802: certrotation: issue one Update request in cert rotation functions

### PR DESCRIPTION
**Pull request by vrutkovs**
Each function used in cert rotation controller should issue just one ApplySecret/ApplySecret per call. This call may  not be atomic, so running several replicas of this controller can cause unexpected races

### CODE REVIEW

File: pkg/operator/certrotation/signer.go
Comment (bertinatto) [inline pkg/operator/certrotation/signer.go:94]: Previously, the code called `applyFn` based on the result of an `&&` condition. Now it uses `||`. Can you confirm if this change is intentional?

The new code looks correct to me, but I just wanted to confirm.

File: pkg/operator/certrotation/target.go
Comment (bertinatto) [inline pkg/operator/certrotation/target.go:128]: Same question WRT `||` and `&&`.

File: pkg/operator/certrotation/signer.go
Comment (bertinatto) [inline pkg/operator/certrotation/signer.go:112]: Could we remove `signerUpdated` now that we have `modified`?

File: pkg/operator/certrotation/signer.go
Comment (vrutkovs) [reply pkg/operator/certrotation/signer.go:94]: Yes. Previously we used `&&` to decide if `Update` needs to be called when either type changed or metadata changed. However, `&&` didn't run `ensureSecretTLSTypeSet` if first part was true, and the type didn't change on first iteration, only the metadata.

By splitting this into separate calls we ensure both `ensureMetadataUpdate` and `ensureSecretTLSTypeSet` are called. Now we can use `||` because either of those will require a separate `Update` call - also, we want `|| modified` there too, as the secret may not exist at all.

Do you think this needs to be a comment to make sure it won't get broken later on?

File: pkg/operator/certrotation/signer.go
Comment (vrutkovs) [reply pkg/operator/certrotation/signer.go:112]: We still need it in `return` - `modified` covers any change to the secret, but `signerUpdate` is a flag for signer contents change (which triggers CA bundle update in `syncWorker`)

File: pkg/operator/certrotation/target.go
Comment (vrutkovs) [reply pkg/operator/certrotation/target.go:128]: Same logic:
* `modified` may be set to true when secret doesn't exist
* both `ensureMetadataUpdate` and `ensureSecretTLSTypeSet` need to run
* if any change was done to the secret it needs to run `Update` to apply it

**Review by bertinatto — COMMENTED** [review]
I'm not fully familiar with this controller, but I reviewed the changes and they appear to be a nice improvement. Tagging the PR, but I'll let the owners approve it.

/lgtm

File: pkg/operator/certrotation/cabundle.go
Comment (dinhxuanvu) [reply pkg/operator/certrotation/cabundle.go:47]: Second this comment.

**Review by dinhxuanvu — COMMENTED** [review]
Looking good. I think it is ready for a lgtm if you can update the variable name to match the others.

File: pkg/operator/certrotation/cabundle.go
Comment (vrutkovs) [reply pkg/operator/certrotation/cabundle.go]: Fixed. `modified` was used in `ApplyConfigMap` call, so I renamed it to `updated`

**Review by dinhxuanvu — APPROVED** [review]
/lgtm

**Review by soltysh — APPROVED** [review]
/approve

**Pull request merged**
